apps/wallet/src/ui/app/redux/slices/sui-objects/Coin.ts

```diff
@@ -155,72 +258,66 @@ export class Coin {
-            return coinID;
-        } else if (balance > amount) {
-            await signer.splitCoin({
-                coinObjectId: coinID,
-                gasBudget: DEFAULT_GAS_BUDGET_FOR_SPLIT,
-                splitAmounts: [Number(balance - amount)],
-            });
-            return coinID;
-        } else {
-            throw new Error(`Insufficient balance`);
+        if (coinWithExactAmount) {
+            return coinWithExactAmount;
         }
-    }
-
-    private static async prepareCoinWithEnoughBalance(
-        signer: RawSigner,
-        coins: SuiMoveObject[],
-        amount: bigint
-    ): Promise<SuiMoveObject> {
-        // Sort coins by balance in an ascending order
-        coins.sort((a, b) =>
-            Coin.getBalance(a) - Coin.getBalance(b) > 0 ? 1 : -1
+        // use transferSui API to get a coin with the exact amount
+        await Coin.transferSui(
+            signer,
+            coins,
+            amount,
+            await signer.getAddress()
         );
 
-        // return the coin with the smallest balance that is greater than or equal to the amount
-        const coinWithSufficientBalance = coins.find(
-            (c) => Coin.getBalance(c) >= amount
+        const coinWithExactAmount2 = await Coin.selectCoinWithExactAmount(
+            signer,
+            coins,
+            amount,
+            true
         );
-        if (coinWithSufficientBalance) {
-            return coinWithSufficientBalance;
+        if (!coinWithExactAmount2) {
+            throw new Error(`requestCoinWithExactAmount failed unexpectedly`);
         }
+        return coinWithExactAmount2;
+    }
 
-        // merge coins to have a coin with sufficient balance
-        // we will start from the coins with the largest balance
-        // and end with the coin with the second smallest balance(i.e., i > 0 instead of i >= 0)
-        // we cannot merge coins with the smallest balance because we
-        // need to have a separate coin to pay for the gas
-        // TODO: there's some edge cases here. e.g., the total balance is enough before spliting/merging
-        // but not enough if we consider the cost of splitting and merging.
-        let primaryCoin = coins[coins.length - 1];
-        for (let i = coins.length - 2; i > 0; i--) {
-            const mergeTxn = await signer.mergeCoin({
-                primaryCoin: Coin.getID(primaryCoin),
-                coinToMerge: Coin.getID(coins[i]),
-                gasBudget: DEFAULT_GAS_BUDGET_FOR_MERGE,
-            });
+    private static async selectCoinWithExactAmount(
+        signer: RawSigner,
+        coins: SuiMoveObject[],
+        amount: bigint,
+        refreshData = false
+    ): Promise<ObjectId | undefined> {
+        const coinsWithSufficientAmount = refreshData
+            ? await signer.provider.selectCoinsWithBalanceGreaterThanOrEqual(
+                  await signer.getAddress(),
+                  amount,
+                  SUI_TYPE_ARG,
```

Comment:
is this only for SUI token? if so, shall we change the function name?

Reply:
Good catch, done!